sql/core/src/main/scala/org/apache/spark/sql/execution/python/RowQueue.scala

```diff
@@ -171,125 +152,59 @@ private[python] case class DiskRowQueue(
  * HybridRowQueue could be safely appended in one thread, and pulled in another thread in the same
  * time.
  */
-private[python] case class HybridRowQueue(
+case class HybridRowQueue(
     memManager: TaskMemoryManager,
     tempDir: File,
     numFields: Int,
     serMgr: SerializerManager)
-  extends MemoryConsumer(memManager, memManager.getTungstenMemoryMode) with RowQueue {
-
-  // Each buffer should have at least one row
-  private var queues = new java.util.LinkedList[RowQueue]()
-
-  private var writing: RowQueue = _
-  private var reading: RowQueue = _
-
-  // exposed for testing
-  private[python] def numQueues(): Int = queues.size()
-
-  def spill(size: Long, trigger: MemoryConsumer): Long = {
-    if (trigger == this) {
-      // When it's triggered by itself, it should write upcoming rows into disk instead of copying
-      // the rows already in the queue.
-      return 0L
-    }
-    var released = 0L
-    synchronized {
-      // poll out all the buffers and add them back in the same order to make sure that the rows
-      // are in correct order.
-      val newQueues = new java.util.LinkedList[RowQueue]()
-      while (!queues.isEmpty) {
-        val queue = queues.remove()
-        val newQueue = if (!queues.isEmpty && queue.isInstanceOf[InMemoryRowQueue]) {
-          val diskQueue = createDiskQueue()
-          var row = queue.remove()
-          while (row != null) {
-            diskQueue.add(row)
-            row = queue.remove()
-          }
-          released += queue.asInstanceOf[InMemoryRowQueue].page.size()
-          queue.close()
-          diskQueue
-        } else {
-          queue
-        }
-        newQueues.add(newQueue)
-      }
-      queues = newQueues
-    }
-    released
-  }
-
-  private def createDiskQueue(): RowQueue = {
-    DiskRowQueue(File.createTempFile("buffer", "", tempDir), numFields, serMgr)
-  }
-
-  private def createNewQueue(required: Long): RowQueue = {
-    val page = try {
-      allocatePage(required)
-    } catch {
-      case _: SparkOutOfMemoryError =>
-        null
-    }
-    val buffer = if (page != null) {
-      new InMemoryRowQueue(page, numFields) {
-        override def close(): Unit = {
-          freePage(this.page)
-        }
-      }
-    } else {
-      createDiskQueue()
-    }
-
-    synchronized {
-      queues.add(buffer)
-    }
-    buffer
-  }
-
-  def add(row: UnsafeRow): Boolean = {
-    if (writing == null || !writing.add(row)) {
-      writing = createNewQueue(4 + row.getSizeInBytes)
-      if (!writing.add(row)) {
-        throw QueryExecutionErrors.failedToPushRowIntoRowQueueError(writing.toString)
-      }
-    }
-    true
-  }
+  extends HybridQueue[UnsafeRow, RowQueue](memManager, tempDir, serMgr) {
 
-  def remove(): UnsafeRow = {
-    var row: UnsafeRow = null
+  override def remove(): UnsafeRow = {
```

Comment:
I'm wondering why this is not in the generic class? Looks like this also can be moved?

Reply:
agree - moved it, thanks!